### PR TITLE
Ipfs companion switch

### PR DESCRIPTION
**Pull request by yrliou**
Fix https://github.com/brave/brave-browser/issues/3918
## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Windows
  - [x] macOS
  - [x] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests && npm run test-security`) on
  - [ ] Windows
  - [x] macOS
  - [x] Linux
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone

## Test Plan:
`npm test brave_browser_tests -- --filter=BraveIPFSExtensionTest.*`

1) Open brave and visit brave://components, make sure IPFS companion extension is not shown.
2) Go to brave://settings and find the extension section to toggle IPFS companion extension, IPFS companion extension will now be installed and shown in brave://components.
3) Toggle again to disable IPFS companion extension and make sure it is not shown in brave://components.
4) Toggle again to enable IPFS companion extension and restart brave, IPFS companion should be shown in brave://components.
5) Toggle again to disable IPFS companion extension and restart brave, IPFS companion should not be shown in brave://components.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

### CODE REVIEW

**Comment by yrliou** [comment]
Not sure why CI result here shows fail because when I checked the result it's all green at https://staging.ci.brave.com/job/brave-browser-build-pr/job/ipfs_switch/1/flowGraphTable/

File: browser/ui/webui/settings/brave_default_extensions_handler.cc [hunk 1]
```diff
@@ -76,3 +81,28 @@ void BraveDefaultExtensionsHandler::SetHangoutsEnabled(
...
+    const base::ListValue* args) {
+  CHECK_EQ(args->GetSize(), 1U);
+  CHECK(profile_);
+  bool value;
+  args->GetBoolean(0, &value);
+
+  extensions::ExtensionService* service =
+    extensions::ExtensionSystem::Get(profile_)->extension_service();
+
+  if (value) {
```
Comment (emerick) [inline browser/ui/webui/settings/brave_default_extensions_handler.cc]: Nit: Really minor, but maybe `enabled` would be a better name for this variable?

File: browser/extensions/brave_component_loader.h [hunk 2]
```diff
@@ -54,6 +54,7 @@ class BraveComponentLoader : public ComponentLoader {
       WILL_REMOVE,
     };
     virtual void OnPdfExtensionAction(PdfExtensionAction action) = 0;
+    virtual void OnComponentRegistered(std::string extension_id) = 0;
```
Comment (emerick) [inline browser/extensions/brave_component_loader.h]: Should this be a const reference just for consistency?

Comment (yrliou) [reply browser/extensions/brave_component_loader.h]: I think it is consistent to `OnComponentRegistered` and `OnComponentReady` defined in this same file, so I'll leave it as is.

File: browser/ui/webui/settings/brave_default_extensions_handler.cc (hunk 1 above)
Comment (yrliou) [reply browser/ui/webui/settings/brave_default_extensions_handler.cc]: resolved by https://github.com/brave/brave-core/pull/2106/commits/ab5ca552427438c729b8564032fe51e2447c2c89

**Comment by yrliou** [comment]
Weird CI result for version mismatch between brave-browser and brave-core, rebased and forced push to give it another try, no code changes.

**Comment by yrliou** [comment]
All green on Mac & windows in the PR build result.
Linux was failing because PR builder seems to have some problem around sync, re-triggered the build with RUN_INIT param manually get around the problem and the result seems fine, so I'm going to merge it.

**Pull request merged**